sips/sip-0xx/sip-0xx-nakamoto.md

```diff
@@ -0,0 +1,188 @@
+- `key_txoff` is the vtxindex for this producer's VRF key registration
+
+The subsequent output(s) in this transaction are the PoX outputs:
+
+1. If the producer set enrollment is in a reward cycle, then outputs 1 through
+   20 must go to the chosen PoX recipients.
+   - Recipients are chosen as described in "Stacking Consensus Algorithm" in
+     SIP-007, using the final block announcement of term N-3 as the source:
+     addresses are chosen without replacement, by using the sortition hash,
+     mixed with the burn header hash of the final block announcement from term
+     N-3 as the seed for the ChaCha12 pseudorandom function to select 20
+     addresses. Since a producer set term lasts 10 Bitcoin blocks, there are 20
+     PoX recipients, 2 per Bitcoin block, to maintain the same number of reward
+     slots and payment frequency.
+   - The order of these outputs does not matter.
+   - Each of these outputs must receive the same amount of BTC.
+   - If the number of remaining addresses in the reward set, N, is less than 20,
+     then the producer set enrollment must burn BTC by including (20-N) burn
+     outputs
+2. Otherwise, the second output must be a burn address.
+
+During a reward cycle, this enrollment transaction will include a somewhat large
+number of outputs: one `OP_RETURN`, twenty stacker rewards, and one change
+address, totaling 22 outputs. While this might seem like a substantial
+transaction, it effectively replaces ten separate transactions under the SIP-007
+leader block commit scheme, each of which would have four outputs (one
+`OP_RETURN`, two stacker rewards, and one change address). Furthermore, the
+enrollment window's duration of ten blocks potentially allows would-be producers
+to take advantage of lower transaction fees during one of those blocks. Despite
+the higher fee for this larger transaction, the cost can be spread out or
+amortized across the ten blocks of the set, resulting in a lower overall cost
+compared to the previous system.
+
+## Censorship Resistance
+
+The producer set enrollments for set _N_ can be included in any of the 10
+Bitcoin blocks in producer set _N-2_. This makes it extremely difficult for a
+Bitcoin miner to censor these transactions, since to do so, they would need to
+control all 10 Bitcoin blocks in that term.
+
+## Selecting Producers
+
+Would-be producers with valid producer set enrollments in term _N-2_ are
+eligible to be included in the producer set for term _N_. The total number of
+producers in a set needs to be limited to prevent coinbase payouts from
+including too many accounts, as this could slow down event processing and even
+open a new DoS vector. This cap will also prevent the block-signing process from
+becoming too expensive. To this end, the total amount of BTC spent in the
+outputs described in "Producer Set Enrollments" is used to select the producer
+set. Would-be producers are ranked by these BTC expenditures, and the top 100
```

Comment:
> Do you think this is necessary to avoid disadvantaging small miners?

I don't think so. Looking at the history of miner contributions in Stacks 2.0, the highest number of miners is ~7-8, so I think it's probably not worth the mechanics in the protocol to support roll overs.

> Does this open a DoS attack?

Depending on implementation, yes. It would need a maximum rollover lifetime and a lower limit.

However, on reflection, I wonder if setting a "top 100" rather than a threshold like 1% of the total is a bad design. A top 100 incentivizes big miners to split up their commitments to crowd out smaller miners. For example:

Reply:
I agree. A 1% minimum is a better option.